### PR TITLE
Isaac Plugin updated for particle rendering with updated Isaac lib

### PR DESCRIPTION
**Pull request by FelixTUD**
1. IsaacPlugin updated for particle rendering
2. Particle sources can be selected in the isaac.param

- [x] depends on https://github.com/ComputationalRadiationPhysics/isaac/pull/89

### CODE REVIEW

File: include/picongpu/plugins/IsaacPlugin.hpp [hunk 1]
```diff
@@ -283,9 +475,214 @@ class IsaacPlugin : public ILightweightPlugin
...
+                    {
+                        double b(json_real_value(json_autopilot));
+                        autopilot = b;
+                    }
+                    else
+                    {
+                        std::cerr << "json error: autopilot needs a double speed!" << std::endl;
+                    }
+                }
+                /*
```
Comment (ax3l) [inline include/picongpu/plugins/IsaacPlugin.hpp]: uh-oh, dead code ;-)

File: include/picongpu/plugins/IsaacPlugin.hpp [hunk 2]
```diff
@@ -283,9 +475,214 @@ class IsaacPlugin : public ILightweightPlugin
...
+                if ( meta && ( json_autopilot = json_object_get(meta, "autopilot") ) )
+                {
+                    if(json_is_real(json_autopilot))
+                    {
+                        double b(json_real_value(json_autopilot));
+                        autopilot = b;
+                    }
+                    else
+                    {
+                        std::cerr << "json error: autopilot needs a double speed!" << std::endl;
```
Comment (ax3l) [inline include/picongpu/plugins/IsaacPlugin.hpp]: "autopilot needs a double speed" does not sound correct :)

File: include/picongpu/plugins/IsaacPlugin.hpp [hunk 3]
```diff
@@ -388,7 +789,11 @@ class IsaacPlugin : public ILightweightPlugin
             };
 
             isaac_for_each_params( sources, SourceInitIterator(), cellDescription, movingWindow );
-
+            isaac_for_each_params( particleSources, ParticleSourceInitIterator(), movingWindow);
+            //particleSources = ParticleList(pSource1);
```
Comment (ax3l) [inline include/picongpu/plugins/IsaacPlugin.hpp]: dead code :)

File: include/picongpu/plugins/IsaacPlugin.hpp [hunk 4]
```diff
@@ -398,12 +803,20 @@ class IsaacPlugin : public ILightweightPlugin
...
+                MovingWindow::getInstance().getWindow( 0 ).globalDimensions.size,
                 subGrid.getLocalDomain().size,
+                subGrid.getLocalDomain().size / SuperCellSize::toRT(),
                 subGrid.getLocalDomain().offset,
+                particleSources,   
                 sources,
                 cellSizeFactor
             );
+            
+            std::cout << "GlobalDomain: " << subGrid.getGlobalDomain().size[0] << "; " << subGrid.getGlobalDomain().size[1] << "; " << subGrid.getGlobalDomain().size[2] <<  std::endl;
```
Comment (ax3l) [inline include/picongpu/plugins/IsaacPlugin.hpp]: Is this debug code?

File: include/picongpu/plugins/IsaacPlugin.hpp [hunk 5]
```diff
@@ -428,7 +841,7 @@ class IsaacPlugin : public ILightweightPlugin
                 particle_count = localNrOfCells * particles::TYPICAL_PARTICLES_PER_CELL * (bmpl::size<VectorAllSpecies>::type::value) * numProc;
                 last_notify = visualization->getTicksUs();
                 if (rank == 0)
-                    log<picLog::INPUT_OUTPUT > ("ISAAC Init succeded");
+                    log<picLog::INPUT_OUTPUT > ("ISAAC + Particle Init succeded");
```
Comment (ax3l) [inline include/picongpu/plugins/IsaacPlugin.hpp]: - typo: succeeded or successful :)

**Comment by PrometheusPi** [comment]
@FelixTUD ComputationalRadiationPhysics/isaac#89 was merged (I checked the tick box above)

**Comment by psychocoderHPC** [comment]
@FelixTUD Could you please rebase this PR against the current dev.

File: include/picongpu/plugins/IsaacPlugin.hpp [hunk 6]
```diff
@@ -205,20 +361,36 @@ struct SourceInitIterator
...
+    // create compile time particle list
+    using ParticleList = bmpl::transform<boost::fusion::result_of::as_list< Particle_Seq >::type,ParticleTransformoperator<bmpl::_1>>::type;
     using VisualizationType = IsaacVisualization
     <
         cupla::AccHost,
         cupla::Acc,
         cupla::AccStream,
         cupla::KernelDim,
         SimDim,
+    ParticleList,
```
Comment (psychocoderHPC) [inline include/picongpu/plugins/IsaacPlugin.hpp]: wrong indention

File: include/picongpu/plugins/IsaacPlugin.hpp [hunk 7]
```diff
@@ -264,6 +436,7 @@ class IsaacPlugin : public ILightweightPlugin
         {
             step = 0;
             bool pause = false;
+            std::ofstream csvFile;
```
Comment (psychocoderHPC) [inline include/picongpu/plugins/IsaacPlugin.hpp]: please remove this

File: include/picongpu/plugins/IsaacPlugin.hpp [hunk 8]
```diff
@@ -283,9 +457,45 @@ class IsaacPlugin : public ILightweightPlugin
                     json_object_set_new( visualization->getJsonMetaRoot(), "cell count", json_integer( cell_count ) );
                     json_object_set_new( visualization->getJsonMetaRoot(), "particle count", json_integer( particle_count ) );
                 }
+
                 uint64_t start = visualization->getTicksUs();
+
+                visualization->kernel_time = 0;
```
Comment (psychocoderHPC) [inline include/picongpu/plugins/IsaacPlugin.hpp]: why do you set `kernel_time` to zero?

File: include/picongpu/plugins/IsaacPlugin.hpp [hunk 9]
```diff
@@ -283,9 +457,45 @@ class IsaacPlugin : public ILightweightPlugin
...
+                        json_array_append(array, json_real(0.0));
+                        json_array_append(array, json_real(autopilot * (double) drawing_time / 1000000.0));
+
+                        json_object_set_new(feedback, "rotation axis", array);
+                        visualization->getCommunicator()->setMessage( feedback );
+
+                    }
+                }
+
+                json_t* json_autopilot;
```
Comment (psychocoderHPC) [inline include/picongpu/plugins/IsaacPlugin.hpp]: What is this? Is it needed for particles?

**Comment by PrometheusPi** [comment]
After rebasing to `dev` and installing everything, I started a default LWFA simulation. The client connected to the server but picongpu never connected to the server. I set the server port and url (Ethernet ip of hemera5 were the server ran)

Any ideas were this error comes from?

**Comment by PrometheusPi** [comment]
As @psychocoderHPC suggested, I tested changing the `libwebsockets` version. When using `libwebsockets` **2.1** instead of **2.4** i got at least a short moment that I could select the running simulation before the server ran into a segfault.

**Comment by PrometheusPi** [comment]
Found a working solution: using `libwebsockets` **2.1-stable** and setting `-DISAAC_JPEG=ON` when building the server worked. However, with `2.4-stable` neither worked.

**Pull request merged**